temporal-sdk/src/main/java/io/temporal/internal/worker/ActivityWorker.java

```diff
@@ -161,44 +160,44 @@ private TaskHandlerImpl(ActivityTaskHandler handler) {
     }
 
     @Override
-    public void handle(PollActivityTaskQueueResponse task) throws Exception {
-
+    public void handle(ActivityTask task) throws Exception {
+      PollActivityTaskQueueResponse r = task.getResponse();
       Scope metricsScope =
           options
               .getMetricsScope()
               .tagged(
                   ImmutableMap.of(
                       MetricsTag.ACTIVITY_TYPE,
-                      task.getActivityType().getName(),
+                      r.getActivityType().getName(),
                       MetricsTag.WORKFLOW_TYPE,
-                      task.getWorkflowType().getName()));
+                      r.getWorkflowType().getName()));
 
       metricsScope
           .timer(MetricsType.ACTIVITY_SCHEDULE_TO_START_LATENCY)
           .record(
               ProtobufTimeUtils.toM3Duration(
-                  task.getStartedTime(), task.getCurrentAttemptScheduledTime()));
+                  r.getStartedTime(), r.getCurrentAttemptScheduledTime()));
 
       // The following tags are for logging.
-      MDC.put(LoggerTag.ACTIVITY_ID, task.getActivityId());
-      MDC.put(LoggerTag.ACTIVITY_TYPE, task.getActivityType().getName());
-      MDC.put(LoggerTag.WORKFLOW_ID, task.getWorkflowExecution().getWorkflowId());
-      MDC.put(LoggerTag.RUN_ID, task.getWorkflowExecution().getRunId());
+      MDC.put(LoggerTag.ACTIVITY_ID, r.getActivityId());
+      MDC.put(LoggerTag.ACTIVITY_TYPE, r.getActivityType().getName());
+      MDC.put(LoggerTag.WORKFLOW_ID, r.getWorkflowExecution().getWorkflowId());
+      MDC.put(LoggerTag.RUN_ID, r.getWorkflowExecution().getRunId());
 
-      propagateContext(task);
+      propagateContext(r);
 
       try {
```

Comment:
Let's move the try to the beginning of the method. For example, if `propagateContext` throws we are going to leak semaphore.

Reply:
Done, we still need to have metric scope and response variables declared outside of the try block though.
I think it should be safe this way.